### ISSUE TITLE
Sample converters capture-req/capture-res don't work. Unable to access captured data.

### ISSUE THREAD

**Issue by StanleyP** [haproxy/haproxy#956]
## Detailed description of the problem

Sample converters capture-req/capture-res don't work.

I declare captures, fetch samples into the captures, but I'm not able to access captured data.

## Expected behavior

I expect that capture-req/capture-res will return content of capture as described in documentation.

## Steps to reproduce the behavior

Start official haproxy 2.2 with provided configuration.

## Do you have any idea what may have caused this?

No.

## Do you have an idea how to solve the issue?

No.

## What is your configuration?

```
global
  master-worker
  nbthread 1

  log stdout format raw local1 info info

  tune.ssl.default-dh-param 2048

defaults
  timeout connect 10s
  timeout client  60s
  timeout server  60s

  log global

  mode http
  option dontlognull
  log-format 'req:"%[always_true,capture-req(0)]" res:"%[always_true,capture-res(0)]"'

frontend http
  bind :80

  declare capture request len 1024
  declare capture response len 1024

  http-request capture req.body id 0
  http-response capture res.body id 0
```

## Output of `haproxy -vv` and `uname -a`

Using official haproxytech/haproxy-debian:2.2 docker image.

**Comment by StanleyP** [comment]
Could you help me please? I still haven't figured out how to use capture-req/capture-res sample converters.

And why capture-req/capture-res are sample converters in the first place? From their description in documentation it seems to me that they should be sample fetches and not converters.

But maybe I am missing something. Examples of usage in the documentation or some sample code on github/web would be really appreciated.

**Comment by capflam** [comment]
very sorry, I thought I answered you. But, no. To capture the request body, you should wait for it. Otherwise, depending on the network or how the client sends the payload, you will only get the headers during analysis. So, add `option http-buffer-request` in your default/frontend section.

But for the response, there is no such option. For now, the most you can do is to use a tcp-response content rule.

**Comment by capflam** [comment]
In my todo-list, I have point to add an http action to wait on condition with a optional timeout. It could be a way to wait for the request or the response body and to replace the `http-buffer-request` option. I must add a feature request to make it visible.

**Comment by StanleyP** [comment]
The option to wait for request/response would be very interesting. For my use case it would suffice until declared capture buffers are filled (or timeout expired possibly). Any idea when this could be available in haproxy?

But ideally I would like to avoid buffering in main haproxy process at all.

Is it possible to have access to every tcp chunk read/written during http transaction?

What I would like is to send every chunk read/written during http transaction to SPOA (pure payload, no distinction between headers and body) a do the buffering, concatenation and parsing in SPOA itself. That way I could measure traffic fragmentation etc. and still dont block main haproxy process.

Can I use tcp-request/tcp-response actions in HTTP proxies?

I already use option http-buffer-request to inspect the request body, but I won't be able to use full request buffering in production. I must find another way definitely.

Does HAPEE offect such functionality?

Sorry to bother you with so many questions, but so far haproxy performs fantastically in our solution/use-case, it is rock-solid, but the full traffic inspection is mandatory requirement so I am trying to find a way to implement it.

**Comment by capflam** [comment]
About the http action to wait the message body, it is one my todo-list but with a low priority. So I don't know how it could be available but probably not soon. But it will do more or less the same than the `http-buffer-request` option.

About the SPOE, the ability to stream the message payload to an agent for analysis or rewrite is on the long term roadmap. But here again, I have no idea when it will be possible to work on it.

> Can I use tcp-request/tcp-response actions in HTTP proxies?

Of course it is possible. Thus it is possible to wait for the response body adding a `tcp-response content` rule with an appropriate `inspect-delay`. But it is less convenient than the `http-buffer-request` option.

> I already use option http-buffer-request to inspect the request body, but I won't be able to use full request buffering in production. I must find another way definitely.

Without the SPOE streaming capabilities, it is not possible. You may eventually try to write your own filter (the SPOE is based on the HAProxy filters API). The HTTP compression filter is a good example.

Because it is not a bug. I will mark it as "work as designed" and close the issue. If you need mode info, Use the mailing list instead (haproxy@formilux.org). Thanks.

**Comment by StanleyP** [comment]
Thank you very much for all the tips and clues, especially usage of tcp-request/tcp-response content!! For some reason I thought that these are only permitted in tcp mode proxies, mea culpa.
With these informations I think I am able to implement what I need.

**Comment by StanleyP** [comment]
Still could not get it working using tcp-request/tcp-response content send-spoe-group.

I created sample configuration I am experimenting with called contentdebug.zip
To run it, just invoke: docker-compose up and make http request using curl (e.g. `curl -d '{}' http://localhost`)

[contentdebug.zip](https://github.com/haproxy/haproxy/files/5589130/contentdebug.zip)

## Configuration

### haproxy.cfg

```
global
	master-worker
	nbthread 1

	log stdout format raw local1 info info

	tune.ssl.default-dh-param 2048

	lua-load /main.lua

defaults
	timeout connect 10s
	timeout client  60s
	timeout server  60s

	log global

	mode http
	option dontlognull
  option httplog

frontend ping
  bind :8080

  http-request set-log-level silent
  http-request use-service lua.ping

frontend http
	bind :80

	filter spoe engine contentdebug config /contentdebug.spoe.conf

	tcp-request inspect-delay 60s
	tcp-request content send-spoe-group contentdebug contentdebug-req-payload

  use_backend api

backend api
	http-request set-log-level silent
  
	filter spoe engine contentdebug config /contentdebug.spoe.conf

	tcp-response inspect-delay 60s
	tcp-response content send-spoe-group contentdebug contentdebug-res-payload

	server api unix@/tmp/api.sock send-proxy

frontend api
	bind unix@/tmp/api.sock accept-proxy

  http-request set-header x-my-reqheader my-reqval
  http-response set-header x-my-resheader my-resval

  use_backend local

backend local
  server local 127.0.0.1:8080

backend contentdebug-spoe
 	mode tcp
 	balance roundrobin
 	timeout connect 5s
 	timeout server  3m
 	server spoe-server 127.0.0.1:12345
```

### contentdebug.spoe.conf

```
[contentdebug]

spoe-agent contentdebug
  groups contentdebug-req-payload contentdebug-res-payload
	option var-prefix  cd
	timeout hello      100ms
	timeout idle       30s
	timeout processing 15ms
	use-backend contentdebug-spoe

spoe-group contentdebug-req-payload
  messages contentdebug-req-payload

spoe-message contentdebug-req-payload
  args id=var(txn.api_unique_id) len=req.len payload=req.payload(0,0)

spoe-group contentdebug-res-payload
  messages contentdebug-res-payload

spoe-message contentdebug-res-payload
  args id=var(txn.api_unique_id) len=res.len payload=res.payload(0,0)

```

## Output

```
$ docker-compose up
Starting haproxy ... done
Starting spoa    ... done
Attaching to haproxy, spoa
haproxy    | Proxy ping started.
haproxy    | Proxy http started.
haproxy    | Proxy api started.
haproxy    | Proxy api started.
haproxy    | Proxy local started.
haproxy    | Proxy contentdebug-spoe started.
haproxy    | [NOTICE] 328/102035 (1) : New worker #1 (6) forked
spoa       | SPOA is listening on port 12345
spoa       | SPOA worker 01 started
spoa       | SPOA worker 02 started
spoa       | SPOA worker 03 started
spoa       | SPOA worker 04 started
spoa       | SPOA worker 05 started
spoa       | contentdebug-req-payload
spoa       |   id: "nil"
spoa       |   len: "0"
spoa       |   payload: ""
haproxy    | 127.0.0.1:57790 [24/Nov/2020:10:20:44.339] api local/local 0/0/0/0/0 200 110 - - ---- 3/1/0/0/0 0/0 "POST / HTTP/1.1"
```

## Result

As you can see only one call to SPOA is made with zero req.len and no payload.

## Expected

All chunks of payload read/written by haproxy sent to SPOA.

I will make a post to mailing list as you suggested, but I wanted to record my results for the record.

**Comment by capflam** [comment]
Don't use `req.payload()` and `res.payload()` from an HTTP proxy because the data are structured and raw data cannot be extracted. I will update the documentation because it is not obvious for now. In addition, I will return an error if these sample fetches are called on an HTTP stream. Sorry about the confusion. Use `req.body` and `res.body` instead.

BTW, you must first wait for data before sending a message to your SPOA. At the TCP level, you should only wait for some data. But you cannot perform exactly what the `http-buffer-request` option do (wait all the payload or the buffer full). In addition, only messages smaller than a buffer (16k by default) will be entirely available. This limitation cannot be risen with the current SPOE design.

**Comment by StanleyP** [comment]
Ok, so I cannot access raw payload from http mode proxies. With req.body/res.body I will only get body and not raw request line and raw headers. My goal is to inspect raw wire content sent to/from haproxy.

Btw, I posted my issue in the mailing list as you suggested

https://www.mail-archive.com/haproxy@formilux.org/msg39041.html

For me the main issue is that the SPOA is called only once by tcp-request content send-spoa-group. I was under the impression that haproxy will reevaluate all acls and tcp-request content rules upon arrival of new chunk of data util inspection delay is expired. Is my assumption wrong? Is tcp-request content send-spoa-group called only once?

I also experimented with other filters and for example trace filter has access to exactly the data I'm interested in (and much more). Would it be feasible to add new type of event to SPOE (e.g. on-request-data, on-response-data)? I am still trying to find solution without touching haproxy source code, but if I have to modify existing filter(s) or implement new one I would very appreciate your reassurance that I go in the right direction.
